docs/man/btlaunchmany.bittornado.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.33.
.TH BTLAUNCHMANY "1" "August 2004" "btlaunchmany (bttornado)" 
.SH NAME
Btlaunchmany \- manual page for btlaunchmany
.SH SYNOPSIS
.B btlaunchmany
\fI<directory> <global options>\fR
.SH DESCRIPTION
<directory> - directory to look for .torrent files (semi-recursive)
.SH OPTIONS
.PP
This program follows the usual GNU command-line syntax, with long options
starting with two dashes ('-'). A summary of options is included below.

.PP
\fB\-\-max_uploads\fR \fInumber\fP
.IP
the maximum \fInumber\fP of uploads to allow at once. (defaults to 7)
.PP
\fB\-\-keepalive_interval\fR \fIseconds\fP
.IP
number of \fIseconds\fP to pause between sending keepalives (defaults to 120.0)
.PP
\fB\-\-download_slice_size\fR \fIbytes\fP
.IP
How many \fIbytes\fP to query for per request. (defaults to 16384)
.PP
\fB\-\-upload_unit_size\fR \fIbytes\fP
.IP
when limiting upload rate, how many \fIbytes\fP to send at a time (defaults to 1460)
.PP
\fB\-\-request_backlog\fR \fInumber\fP
.IP
maximum \fInumber\fP of requests to keep in a single pipe at once. (defaults to 10)
.PP
\fB\-\-max_message_length\fR \fIlength\fP
.IP
maximum \fIlength\fP prefix encoding you'll accept over the wire - larger values get the
connection dropped. (defaults to 8388608)
.PP
\fB\-\-ip\fR \fIip\fP
.IP
\fIip\fP to report you have to the tracker. (defaults to '')
.PP
\fB\-\-minport\fR \fIportnum\fP
.IP
set \fIportnum\fP as the minimum port to listen on, counts up if unavailable (defaults to 10000)
.PP
\fB\-\-maxport\fR \fIportnum\fP
.IP
set \fIportnum\fP as the maximum port to listen on (defaults to 60000)
.PP
\fB\-\-random_port\fR \fI 0 | 1 \fP
.IP
whether to choose randomly inside the port range instead of counting up linearly
(defaults to 1)
.PP
\fB\-\-responsefile\fR \fIfile\fP
.IP
\fIfile\fP the server response was stored in, alternative to url (defaults to '')
.PP
\fB\-\-url\fR \fIURL\fP
.IP
\fIURL\fP to get file from, alternative to responsefile (defaults to '')
.PP
\fB\-\-selector_enabled\fR \fI 0 | 1 \fP
.IP
whether to enable the file selector and fast resume function (defaults to 1)
.PP
\fB\-\-expire_cache_data\fR \fIdays\fP
.IP
the number of \fIdays\fP after which you wish to expire old cache data (0 = disabled) (defaults
to 10)
.PP
\fB\-\-priority\fR \fI -1|0|1|2[,-1|0|1|2] \fP
.IP
a list of file priorities separated by commas, must be one per file, 0 = highest, 1 =
normal, 2 = lowest, \fB\-1\fR = download disabled (defaults to ''). Order is based
on the file/torrent order as shown by btshowmetainfo. For example, to download only
the third of four files use: --priority -1,-1,2,-1
.PP
\fB\-\-saveas\fR \fIfilename\fP
.IP
local \fIfilename\fP to save the file as, null indicates query user (defaults to '')
.PP
\fB\-\-timeout\fR \fIseconds\fP
.IP
time in \fIseconds\fP to wait between closing sockets which nothing has been received on (defaults to
300.0)
.PP
\fB\-\-timeout_check_interval\fR \fIseconds\fP
.IP
time to wait in \fIseconds\fP between checking if any connections have timed out (defaults to 60.0)
.PP
\fB\-\-max_slice_length\fR \fIlength\fP
.IP
maximum \fIlength\fP slice to send to peers, larger requests are ignored (defaults to 131072)
.PP
\fB\-\-max_rate_period\fR \fIseconds\fP
.IP
maximum amount of time in \fIseconds\fP to guess the current rate estimate represents (defaults to 20.0)
.PP
\fB\-\-bind\fR \fIip[,hostname]\fP
.IP
comma-separated list of \fIips/hostnames\fP to bind to locally (defaults to '')
.PP
\fB\-\-ipv6_enabled\fR \fI 0 | 1 \fP
.IP
allow the client to connect to peers via IPv6 (defaults to 0)
.PP
\fB\-\-ipv6_binds_v4\fR \fI 0 | 1 \fP
.IP
set if an IPv6 server socket will also field IPv4 connections (defaults to 1)
.PP
\fB\-\-upnp_nat_access\fR \fI 0 | 1 | 2 \fP
.IP
attempt to autoconfigure a UPnP router to forward a server port (0 = disabled, 1 = mode 1
[fast], 2 = mode 2 [slow]) (defaults to 1)
.PP
\fB\-\-upload_rate_fudge\fR \fIseconds\fP
.IP
time equivalent in \fIseconds\fP of writing to kernel-level TCP buffer, for rate adjustment (defaults to
5.0)
.PP
\fB\-\-tcp_ack_fudge\fR \fIoverhead\fP
.IP
how much TCP ACK download \fIoverhead\fP to add to upload rate calculations (0 = disabled)
(defaults to 0.029999999999999999)
.PP
\fB\-\-display_interval\fR \fIseconds\fP
.IP
time in \fIseconds\fP between updates of displayed information (defaults to 0.5)
.PP
\fB\-\-rerequest_interval\fR \fIseconds\fP
.IP
time to wait, in \fIseconds\fP, between requesting more peers (defaults to 300)
.PP
\fB\-\-min_peers\fR \fInumber\fP
.IP
minimum \fInumber\fP of peers to not do rerequesting (defaults to 20)
.PP
\fB\-\-http_timeout\fR \fIseconds\fP
.IP
number of \fIseconds\fP to wait before assuming that an http connection has timed out (defaults
to 60)
.PP
\fB\-\-max_initiate\fR \fInumber\fP
.IP
\fInumber\fP of peers at which to stop initiating new connections (defaults to 40)
.PP
\fB\-\-check_hashes\fR \fI 0 | 1 \fP
.IP
whether to check hashes on disk (defaults to 1)
.PP
\fB\-\-max_upload_rate\fR \fIkB/s\fP
.IP
maximum \fIkB/s\fP to upload at (0 = no limit, \fB\-1\fR = automatic) (defaults to 0)
.PP
\fB\-\-max_download_rate\fR \fIkB/s\fP
.IP
maximum \fIkB/s\fP to download at (0 = no limit) (defaults to 0)
.PP
\fB\-\-alloc_type\fR \fInormal | background | pre-allocate | sparse\fP
.IP
allocation type (may be normal, background, pre-allocate or sparse) (defaults to
\&'normal')
.PP
\fB\-\-alloc_rate\fR \fIMiB/s\fP
.IP
rate (in \fIMiB/s\fP) to allocate space at using background allocation (defaults to 2.0)
.PP
\fB\-\-buffer_reads\fR \fI 0 | 1 \fP
.IP
whether to buffer disk reads (defaults to 1)
.PP
\fB\-\-write_buffer_size\fR \fIspace\fP
.IP
the maximum amount of \fIspace\fP to use for buffering disk writes (in megabytes, 0 = disabled)
(defaults to 4)
.PP
\fB\-\-snub_time\fR \fIseconds\fP
.IP
\fIseconds\fP to wait for data to come in over a connection before assuming it's
semi-permanently choked (defaults to 30.0)
.PP
\fB\-\-spew\fR \fI 0 | 1 \fP
.IP
whether to display diagnostic info to stdout (defaults to 0)
.PP
\fB\-\-rarest_first_cutoff\fR \fInumber\fP
.IP
\fInumber\fP of downloads at which to switch from random to rarest first (defaults to 2)
.PP
\fB\-\-rarest_first_priority_cutoff\fR \fInumber\fP
.IP
the \fInumber\fP of peers which need to have a piece before other partials take priority over
rarest first (defaults to 5)
.PP
\fB\-\-min_uploads\fR \fInumber\fP
.IP
the \fInumber\fP of uploads to fill out to with extra optimistic unchokes (defaults to 4)
.PP
\fB\-\-max_files_open\fR \fInumber\fP
.IP
the maximum \fInumber\fP of files to keep open at a time, 0 means no limit (defaults to 50)
.PP
\fB\-\-round_robin_period\fR \fIseconds\fP
.IP
the number of \fIseconds\fP between the client's switching upload targets (defaults to 30)
.PP
\fB\-\-super_seeder\fR \fI 0 | 1 \fP
.IP
whether to use special upload-efficiency-maximizing routines (only for dedicated seeds)
(defaults to 0)
.PP
\fB\-\-security\fR \fI 0 | 1 \fP
.IP
whether to enable extra security features intended to prevent abuse (defaults to 1)
.PP
\fB\-\-max_connections\fR \fInumber\fP
.IP
the absolute maximum \fInumber\fP of peers to connect with (0 = no limit) (defaults to 0)
.PP
\fB\-\-auto_kick\fR \fI 0 | 1 \fP
.IP
whether to allow the client to automatically kick/ban peers that send bad data (defaults
to 1)
.PP
\fB\-\-double_check\fR \fI 0 | 1 \fP
.IP
whether to double-check data being written to the disk for errors (may increase CPU load)
(defaults to 1)
.PP
\fB\-\-triple_check\fR \fI 0 | 1 \fP
.IP
whether to thoroughly check data being written to the disk (may slow disk access)
(defaults to 0)
.PP
\fB\-\-lock_files\fR \fI 0 | 1 \fP
.IP
whether to lock files the client is working with (defaults to 1)
.PP
\fB\-\-lock_while_reading\fR \fI 0 | 1 \fP
.IP
whether to lock access to files being read (defaults to 0)
.PP
\fB\-\-auto_flush\fR \fIminutes\fP
.IP
\fIminutes\fP between automatic flushes to disk (0 = disabled) (defaults to 0)
.PP
\fB\-\-parse_dir_interval\fR \fIseconds\fP
.IP
how often to rescan the torrent directory, in \fIseconds\fP (defaults to
60)
.PP
\fB\-\-saveas_style\fR \fI 1 | 2 | 3 \fP
.IP
How to name torrent downloads (1 = rename to torrent name, 2 = save
under name in torrent, 3 = save in directory under torrent name)
(defaults to 1)
.PP
\fB\-\-display_path\fR \fI 0 | 1 \fP
.IP
whether to display the full path or the torrent contents for each
torrent (defaults to 1)
.PP
\fB\-\-save_options\fR \fI 0 | 1 \fP
.IP
whether to save the current options as the new default configuration
(only for btlaunchmany) (defaults to 0)
.PP
.SH SEE ALSO
.BR bittorrent-downloader (1),
.BR btmakemetafile (1).
.br
.SH AUTHOR
This manual page was written by Micah Anderson <micah@riseup.net>,
for the Debian GNU/Linux system (but may be used by others).


btlaunchmany.py
#!/usr/bin/env python

# Written by John Hoffman
# see LICENSE.txt for license information

from BitTornado import PSYCO
if PSYCO.psyco:
    try:
        import psyco
        assert psyco.__version__ >= 0x010100f0
        psyco.full()
    except:
        pass

from BitTornado.launchmanycore import LaunchMany
from BitTornado.download_bt1 import defaults, get_usage
from BitTornado.parseargs import parseargs
from threading import Event
from sys import argv, exit
import sys, os
from BitTornado import version, report_email
from BitTornado.ConfigDir import ConfigDir

assert sys.version >= '2', "Install Python 2.0 or greater"
try:
    True
except:
    True = 1
    False = 0

def hours(n):
    if n == 0:
        return 'complete!'
    try:
        n = int(n)
        assert n >= 0 and n < 5184000  # 60 days
    except:
        return '<unknown>'
    m, s = divmod(n, 60)
    h, m = divmod(m, 60)
    if h > 0:
        return '%d hour %02d min %02d sec' % (h, m, s)
    else:
        return '%d min %02d sec' % (m, s)


Exceptions = []

class HeadlessDisplayer:
    def display(self, data):
        print ''
        if not data:
            self.message('no torrents')
        for x in data:
            ( name, status, progress, peers, seeds, seedsmsg, dist,
              uprate, dnrate, upamt, dnamt, size, t, msg ) = x
            print '"%s": "%s" (%s) - %sP%s%s%.3fD u%0.1fK/s-d%0.1fK/s u%dK-d%dK "%s"' % (
                        name, status, progress, peers, seeds, seedsmsg, dist,
                        uprate/1000, dnrate/1000, upamt/1024, dnamt/1024, msg)
        return False
            
    def message(self, s):
        print "### "+s

    def exception(self, s):
        Exceptions.append(s)
        self.message('SYSTEM ERROR - EXCEPTION GENERATED')


if __name__ == '__main__':
    if argv[1:] == ['--version']:
        print version
        exit(0)
    defaults.extend( [
        ( 'parse_dir_interval', 60,
          "how often to rescan the torrent directory, in seconds" ),
        ( 'saveas_style', 1,
          "How to name torrent downloads (1 = rename to torrent name, " +
          "2 = save under name in torrent, 3 = save in directory under torrent name)" ),
        ( 'display_path', 1,
          "whether to display the full path or the torrent contents for each torrent" ),
    ] )
    try:
        configdir = ConfigDir('launchmany')
        defaultsToIgnore = ['responsefile', 'url', 'priority']
        configdir.setDefaults(defaults,defaultsToIgnore)
        configdefaults = configdir.loadConfig()
        defaults.append(('save_options',0,
         "whether to save the current options as the new default configuration " +
         "(only for btlaunchmany.py)"))
        if len(argv) < 2:
            print "Usage: btlaunchmany.py <directory> <global options>\n"
            print "<directory> - directory to look for .torrent files (semi-recursive)"
            print get_usage(defaults, 80, configdefaults)
            exit(1)
        config, args = parseargs(argv[1:], defaults, 1, 1, configdefaults)
        if config['save_options']:
            configdir.saveConfig(config)
        configdir.deleteOldCacheData(config['expire_cache_data'])
        if not os.path.isdir(args[0]):
            raise ValueError("Warning: "+args[0]+" is not a directory")
        config['torrent_dir'] = args[0]
    except ValueError, e:
        print 'error: ' + str(e) + '\nrun with no args for parameter explanations'
        exit(1)

    LaunchMany(config, HeadlessDisplayer())
    if Exceptions:
        print '\nEXCEPTION:'
        print Exceptions[0]
        print 'please report this to '+report_email